octobot/community/authentication.py

```diff
@@ -0,0 +1,177 @@
+#
+#  You should have received a copy of the GNU Lesser General Public
+#  License along with this library.
+import functools
+import time
+import requests
+
+import octobot.constants as constants
+import octobot_commons.constants as commons_constants
+import octobot_commons.config_manager as config_manager
+import octobot_commons.logging as bot_logging
+
+
+def authenticated(func):
+    @functools.wraps(func)
+    def wrapped(self, *args, **kwargs):
+        self.ensure_token_validity()
+        return func(self, *args, **kwargs)
+    return wrapped
+
+
+class CommunityAuthentication:
+    """
+    Authentication utility
+    """
+    ALLOWED_TIME_DELAY = 1 * commons_constants.MINUTE_TO_SECONDS
+    AUTHORIZATION_HEADER = "Authorization"
+
+    def __init__(self, authentication_url, username=None, password=None, config=None):
+        self.logger = bot_logging.get_logger(self.__class__.__name__)
+        self.authentication_url = authentication_url
+        self.refresh_token = None
+        self.edited_config = config
+
+        self._auth_token = None
+        self._expire_at = None
+        self._session = requests.Session()
+        self._cache = {}
+
+        if None not in (username, password):
+            self.login(username, password)
+
+    def get_logged_in_email(self):
+        return self.get(constants.OCTOBOT_COMMUNITY_ACCOUNT_URL).json()["data"]["attributes"]["email"]
+
+    def login(self, username, password):
+        self._reset_tokens()
+        params = {
+            "username": username,
+            "password": password,
+            "grant_type": "password",
+        }
+        resp = requests.post(self.authentication_url, params=params)
+        self._handle_auth_result(resp)
+
+    def logout(self):
+        self._reset_tokens()
+        self.clear_cache()
+        self.remove_login_detail()
+
+    def clear_cache(self):
+        self._cache = {}
+
+    @authenticated
+    def get(self, url, params=None, allow_cache=False, **kwargs):
+        if allow_cache:
+            if url not in self._cache:
+                self._cache[url] = self._session.get(url, params=params, **kwargs)
+            return self._cache[url]
+        else:
+            return self._session.get(url, params=params, **kwargs)
+
+    @authenticated
+    def post(self, url, data=None, json=None, **kwargs):
+        return self._session.post(url, data=data, json=json, **kwargs)
+
+    def is_logged_in(self):
+        return None not in (self._auth_token, self.refresh_token, self._expire_at)
+
+    def ensure_token_validity(self):
+        if not self.is_logged_in():
+            # try to login with saved token if available
+            self._try_auto_login()
+            if not self.is_logged_in():
+                # still not logged in: raise
+                raise AuthenticationRequired()
+        if time.time() >= self._expire_at:
+            self._refresh_auth()
+
+    def remove_login_detail(self):
+        self._update_login_config("")
```

Comment:
Why not passing no argument with `_update_login_config(self, value="")` ?

Reply:
Here `_update_login_config` is supposed to receive a value, "" is more of an exception than the default behavior

Reply:
Ok, in my opinion it's not very clear to say "update something with an empty value" to reset / remove a thing.

Reply:
Right, I'll see how to make it clearer